pkg/network/protocols/http/shared_libraries.go

```diff
@@ -160,221 +257,58 @@ func (w *soWatcher) Start() {
-	if w.hostMount != "" {
-		path = filepath.Join(w.hostMount, path)
-	}
-
-	return followSymlink(path)
-}
-
-type soRegistration struct {
-	// set of pid paths (eg. /host/proc/123) that are used for liveness checks
-	// this is done to unregister shared libraries as soon as all processes
-	// referencing it terminate. This is required when the shared library is
-	// located in a volume where mount propagation is enabled, so containers
-	// don't get stuck during termination.
-	pids map[string]struct{}
-
-	inode        uint64
-	refcount     int
-	unregisterCB func(string) error
-}
-
-type soRegistry struct {
-	procRoot string
-	byPath   map[string]*soRegistration
-	byInode  map[uint64]*soRegistration
-}
-
-func (r *soRegistry) register(libPath string, pid int, rule soRule) {
-	pidPath := fmt.Sprintf("%s/%d", r.procRoot, pid)
-	if registration, ok := r.byPath[libPath]; ok {
-		registration.pids[pidPath] = struct{}{}
+func (r *soRegistry) Register(root string, libPath string, pid uint32, rule soRule) {
+	hostLibPath := root + libPath
+	pathID, err := newPathIdentifier(hostLibPath)
+	if err != nil {
+		// short living process can't hit here
+		// as we receive the openat() syscall info after receiving the EXIT netlink process
+		log.Debugf("can't create path identifier %s", err)
 		return
 	}
-
-	inode, err := getInode(libPath)
-	if err != nil {
+	if _, found := r.blocklistByID[pathID]; found {
 		return
 	}
 
-	if registration, ok := r.byInode[inode]; ok {
-		registration.refcount++
-		r.byPath[libPath] = registration
-		registration.pids[pidPath] = struct{}{}
-		log.Debugf("registering library=%s", libPath)
+	r.m.Lock()
+	defer r.m.Unlock()
+	if reg, found := r.byID[pathID]; found {
+		reg.refcount++
+		r.byPID[pid] = reg
 		return
 	}
 
-	err = rule.registerCB(libPath)
-	if err != nil {
-		log.Debugf("error registering library=%s: %s", libPath, err)
-		if err := rule.unregisterCB(libPath); err != nil {
-			log.Debugf("unregisterCB %s : %s", libPath, err)
+	if err := rule.registerCB(pathID, root, libPath); err != nil {
+		log.Debugf("error registering library %s path %s by pid %d : %s", pathID.String(), hostLibPath, pid, err)
+		if rule.unregisterCB != nil {
+			if err := rule.unregisterCB(pathID); err != nil {
+				log.Debugf("unregisterCB library %s path %s : %s", pathID.String(), hostLibPath, err)
+			}
 		}
-
 		// save sentinel value so we don't attempt to re-register shared
 		// libraries that are problematic for some reason
-		registration := newRegistration(inode, nil)
-		r.byPath[libPath] = registration
-		r.byInode[inode] = registration
-		return
-	}
-
-	log.Debugf("registering library=%s", libPath)
-	registration := newRegistration(inode, rule.unregisterCB)
-	registration.pids[pidPath] = struct{}{}
-	r.byPath[libPath] = registration
-	r.byInode[inode] = registration
-}
-
-func (r *soRegistry) unregister(libPath string) {
-	registration, ok := r.byPath[libPath]
-	if !ok {
+		r.blocklistByID[pathID] = struct{}{}
```

Comment:
that's is an always increasing memory

Reply:
yes but only on problematic path, so this scope is very limited
The only way to cleanup this entry would be the mTime check of the pathID, see jira USMO-172